tests/test_mutations.py

```diff
@@ -472,40 +704,226 @@ def test_keep_mutation_parent_zero_rate(self):
+                print(f"mut {j}\t{mut_item['time']}\t{mut_item['new']}")
+                print(mut_item['mutation'])
+
+    def generate(self, tables, seed, keep=False, discrete_sites=False):
+        rng = msprime.RandomGenerator(seed)
+        if keep:
+            self.initialise_sites(tables)
+        tables.sites.clear()
+        tables.mutations.clear()
+        self.place_mutations(tables, rng, discrete_sites=discrete_sites)
+        self.populate_tables(tables)
+        tables.compute_mutation_parents()
+        self.record_provenance(tables, seed, keep, discrete_sites)
+        self.choose_alleles(tables, rng)
+        return tables.tree_sequence()
+
+    def initialise_sites(self, tables):
+        sites = tables.sites
+        mutations = tables.mutations
+        nodes = tables.nodes
+        j = 0
+        for site_id in range(sites.num_rows):
+            site = sites[site_id]
+            mutation_list = []
+            while j < mutations.num_rows and mutations.site[j] == site_id:
+                mut = mutations[j]
+                mutation_item = {'time': nodes.time[mut.node],
+                                 'mutation': mut,
+                                 'new': False}
+                mutation_list.append(mutation_item)
+                j += 1
+
+            mutation_list.reverse()
+            site_item = {'site': site,
+                         'mutation_list': mutation_list}
+            self.sites.append(site_item)
+
+    def populate_tables(self, tables):
+        new_mutations = 0
+        for site_item in self.sites:
+            site = site_item['site']
+            site_id = tables.sites.add_row(**site._asdict())
+            for m in site_item['mutation_list']:
+                mut = m['mutation']
+                parent = mut.parent
+                if parent != tskit.NULL:
+                    parent += new_mutations
+                tables.mutations.add_row(site_id, mut.node, mut.derived_state,
+                                         parent=parent, metadata=mut.metadata)
+                if m['new']:
+                    new_mutations += 1
+
+    def place_mutations(self, tables, rng, discrete_sites=False):
+        # Insert a sentinel into the map for convenience.
+        map_position = np.hstack([self.rate_map.position, [tables.sequence_length]])
+        node_times = tables.nodes.time
+        for edge in tables.edges:
+            branch_start = node_times[edge.child]
+            branch_end = node_times[edge.parent]
+            branch_length = branch_end - branch_start
+            index = np.searchsorted(map_position, edge.left)
+            if map_position[index] > edge.left:
+                index -= 1
+            left = edge.left
+            right = 0
+            while right != edge.right:
+                right = min(edge.right, map_position[index + 1])
+                site_left = np.ceil(left) if discrete_sites else left
```

Comment:
So  yes, I think we just need to drop the ceiling here to compute the rates correctly.